articles/logging-with-winston/index.md

```diff
@@ -0,0 +1,576 @@
+
+Winston allows you to implement [multiple logging transports](https://github.com/winstonjs/winston#working-with-multiple-loggers-in-winston), i.e., a log can be recorded to a file, console, or database.
+
+The Logger configuration below logs to a console and a file. We will add a transport array to the log configuration object. Later in this guide, we will show you how to log and record the log instances to a database.
+
+```js
+const logConfiguration = {
+    transports: [
+        new winston.transports.Console({
+            level: 'warn'
+        }),
+        new winston.transports.File({
+            level: 'error',
+            // Create the log directory if it does not exist
+            filename: './logs/example.log'
+        })
+    ]
+};
+
+const logger = winston.createLogger(logConfiguration);
+
+// Log some messages
+logger.error("Hello, Winston logger, the first error!");
+logger.warn("Hello, Winston logger, the first warning!");
+logger.warn("Hello, Winston logger, the second warning!");
+logger.error("Hello, Winston logger, the second error!");
+logger.info("Hello, Winston logger, some info!");
+logger.debug("Hello, Winston logger, a debug!");
+```
+
+This specifies that:
+
+- Logs will be shown in the console output.
+- Only the log that falls under level `error` will be recorded in the `example.log` file.
+
+With Winston, you can specify the default format to save your logs.
+
+For example, let's say we want to log in JSON format we can specify that with `Winston.format`, and log instances will be saved in JSON format.
+
+```js
+format: winston.format(
+)
+```
+
+The format takes additional log form properties such as
+
+- `ms()` - time in milliseconds since the last log was recorded.
+- `label()` - adds a label to the logged message.
+- `timestamp()` - the time stamp the log message was received.
+- `splat()` - provides string interpolation.
+
+To apply this in your logs, you need to use `format.combine`, as shown in the example below.
+
+```js
+const logConfiguration = {
+    transports: [
+        new winston.transports.Console()
+    ],
+    format: winston.format.combine(
+        winston.format.label({
+            label: `Label🏷️`
+        }),
+        winston.format.timestamp({
+           format: 'MMM-DD-YYYY HH:mm:ss'
+       }),
+        winston.format.printf(info => `${info.level}: ${info.label}: ${[info.timestamp]}: ${info.message}`),
+    )
+};
+
+const logger = winston.createLogger(logConfiguration);
+
+// Log a message
+logger.info("Hello, Winston logger, some info!");
+```
+
+Output:
+
+```js
+info: Label🏷️: Nov-18-2020 08:10:44: Hello, Winston logger, some info!
+```
+
+### Configuring Winston with a Server
+Let’s create a simple Express server that we can do some logging using Winston. This will be a small project to get you on your feet using Winston to record logs from server requests and responses.
+
+Go ahead and install the [Express](/engineering-education/express/) library with `npm install express`.
```

Comment:
Need to install winston-mongodb here rather than later since the module is required in the below code snippet.